lib/src/message.dart

```diff
@@ -149,55 +149,89 @@ abstract class Message {
-    } else {
-      var contentType = new MediaType.parse(newHeaders['content-type'])
-          .change(parameters: {'charset': body.encoding.name});
-      newHeaders['content-type'] = contentType.toString();
+    if (contentType.isNotEmpty) {
+      newHeaders['content-type'] = contentType;
     }
-  }
 
-  if (body.contentLength != null) {
-    var coding = newHeaders['transfer-encoding'];
-    if (coding == null || equalsIgnoreAsciiCase(coding, 'identity')) {
-      newHeaders['content-length'] = body.contentLength.toString();
+    if (contentLength.isNotEmpty) {
+      newHeaders['content-length'] = contentLength;
     }
+
+    return newHeaders;
   }
 
-  return newHeaders;
-}
+  /// Determines the `content-length` from the given [headers] and [body].
+  ///
+  /// Returns the value for the `content-length` header if it should be
+  /// modified, otherwise it returns the empty string.
+  static String _contentLengthHeader(Map<String, String> headers, Body body) {
+    var bodyLength = body.contentLength;
+    if (bodyLength == null) return '';
+
+    var contentLengthHeader = bodyLength.toString();
+    if (contentLengthHeader == getHeader(headers, 'content-length')) return '';
 
-/// Returns whether [headers] declares the same encoding as [body].
-bool _sameEncoding(Map<String, String> headers, Body body) {
-  if (body.encoding == null) return true;
+    var coding = getHeader(headers, 'transfer-encoding');
+
+    return coding == null || equalsIgnoreAsciiCase(coding, 'identity')
+        ? contentLengthHeader
+        : '';
+  }
 
-  var contentType = getHeader(headers, 'content-type');
-  if (contentType == null) return false;
+  /// The default media type `application/octet-stream` as defined by HTTP.
+  static final MediaType _defaultMediaType =
+      new MediaType('application', 'octet-stream');
 
-  var charset = new MediaType.parse(contentType).parameters['charset'];
-  return Encoding.getByName(charset) == body.encoding;
+  /// Determines the `content-type` from the given [headers] and [body].
+  ///
+  /// The function looks at the encoding of the body and encoding specified
+  /// within the `content-type` header. The body's encoding will always
+  /// override the value.
+  ///
+  /// Returns the value for the `content-type` header if it should be
+  /// modified, otherwise it returns the empty string.
+  static String _contentTypeHeader(Map<String, String> headers, Body body) {
+    var contentTypeHeader = getHeader(headers, 'content-type');
+    var changed = false;
+    MediaType mediaType;
+    Encoding mediaEncoding = null;
+
+    if (contentTypeHeader != null) {
+      mediaType = new MediaType.parse(contentTypeHeader);
+      mediaEncoding = Encoding.getByName(mediaType.parameters['charset']);
+    } else {
+      mediaType = _defaultMediaType;
+    }
+
+    var bodyEncoding = body.encoding;
+
+    if ((bodyEncoding != null) && (bodyEncoding != mediaEncoding)) {
+      changed = true;
```

Comment:
Rather than tracking `changed`, I'd just write:

```dart
if (body.encoding == null) return null;
if (body.encoding == mediaEncoding) return null;

return mediaType.change(...);
```